concordium-consensus/tests/scheduler/SchedulerTests/Payday.hs

```diff
@@ -39,70 +41,82 @@ import Concordium.GlobalState.Basic.BlockState.PoolRewards (BakerPoolRewardDetai
-    it "two updates" $ assertEqual "" (runMock (events mintAmts2) (op [(0, UVMintDistribution md1), (100, UVMintDistribution md2)])) result
-    it "two updates, one late" $ assertEqual "" (runMock (events mintAmts1) (op [(0, UVMintDistribution md1), (401, UVMintDistribution md2)])) result
+    it "no updates" $
+        runTest [] [mintAmts0]
+    it "last minute update" $
+        runTest [(400, UVMintDistribution md1)] [mintAmts1]
+    it "late update" $
+        runTest [(401, UVMintDistribution md1)] [mintAmts0]
+    it "two updates" $
+        runTest [(0, UVMintDistribution md1), (100, UVMintDistribution md2)] [mintAmts2]
+    it "two updates, one late" $
+        runTest [(0, UVMintDistribution md1), (401, UVMintDistribution md2)] [mintAmts1]
     -- Note: the minting should not be affected by any updates to the time parameters
-    it "mint rate update" $ assertEqual "" (runMock (events mintAmts0) (op [(0, UVTimeParameters (TimeParametersV1 100 (MintRate 2 0)))])) result
+    it "mint rate update" $
+        runTest [(0, UVTimeParameters (TimeParametersV1 100 (MintRate 2 0)))] [mintAmts0]
   where
-    events :: MintAmounts -> [WithResult (Action 'P4)]
-    events amts =
-        [ BSO (BsoGetBankStatus (bs 0)) :-> bank,
-          BSO (BsoGetSeedState (bs 0)) :-> seedState,
-          BSO (BsoMint (bs 0) amts) :-> bs 1,
-          BSO (BsoAddSpecialTransactionOutcome (bs 1) (mintSto amts)) :-> bs 2
-        ]
-    op upds = doMintingP4 dummyChainParameters targetEpoch mintRate foundationAccount upds (bs 0)
-    result = bs 2
-    bs = MockUpdatableBlockState
+    initialTotalSupply = 1_000_000_000
+    epochLength = 100
+    -- Initial block state, the important information here is the epoch length and the initial total
+    -- supply.
+    initialBlockState = do
+        account <- Helpers.makeTestAccountFromSeed initialTotalSupply 0
+        BS.initialPersistentState
+            (initialSeedState (Hash.hash "NONCE") epochLength)
+            DummyData.dummyCryptographicParameters
+            [account]
+            DummyData.dummyIdentityProviders
+            DummyData.dummyArs
+            DummyData.dummyKeyCollection
+            DummyData.dummyChainParameters
+    -- Run a test of doMintingP4. It is provided a list of updates (paired with effective slot time)
+    -- and a list of expected special transaction outcomes to have been produced.
+    runTest :: [(Slot, UpdateValue 'ChainParametersV1)] -> [SpecialTransactionOutcome] -> Assertion
+    runTest updates expectedSpecialOutcomes = join $ Helpers.runTestBlockState $ do
+        initialState :: PersistentBlockState 'P4 <- thawBlockState =<< initialBlockState
+        newState <- doMintingP4 dummyChainParameters targetEpoch mintRate foundationAccount updates initialState
+        specialOutcomes <- getSpecialOutcomes =<< hashBlockState newState
+        return $ assertEqual "Incorrect special outcomes are produced" expectedSpecialOutcomes $ toList specialOutcomes
```

Comment:
We should not just be checking the `SpecialTransactionOutcome`s, but we should also check that the correct amount was actually minted in the state. Ideally, we should also test that no other changes were made to the state, but that might be a bit tricky to implement.